### PR TITLE
Add initial support for UltraDNS provider

### PR DESCRIPTION
**Pull request by abligh**
Signed-off-by: Alex Bligh <alex@alex.org.uk>

This lightly tested PR adds support for ultradns. It passes all tests, plus some manual ones, but I have not yet used it in anger.

### CODE REVIEW

File: lexicon/tests/providers/test_ultradns.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+class UltradnsProviderTests(TestCase, IntegrationTests):
+    """TestCase for UltraDNS"""
+    provider_name = 'ultradns'
+    domain = 'example-abtest.com'
+
+    def _filter_headers(self):
+        return ['Authorization']
+
+    def _filter_post_data_parameters(self):
+        return ['username', 'password']
```
Comment (adferrand) [inline lexicon/tests/providers/test_ultradns.py:19]: It seems that parameter `accessToken` in the returned JSON during an authentication would also be useful to be filtered, if someone refresh the data, but using a user+password credentials this time.

File: lexicon/providers/ultradns.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+LOGGER = logging.getLogger(__name__)
+
+NAMESERVER_DOMAINS = ['ultradns.net']
+
+
+def provider_parser(subparser):
+    """Configure provider parser for ultradns"""
+    subparser.add_argument(
+        "--auth-token", help="specify token for authentication")
```
[the remark this answers is not in the document]
Comment (abligh) [reply lexicon/providers/ultradns.py]: Fixed

File: lexicon/providers/ultradns.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        super(Provider, self).__init__(config)
+
+        self.api_key = self._get_provider_option('auth_token')
+
+        assert (self.api_key is not None) or \
+            (self._get_provider_option('auth_username') is not None and \
+                 self._get_provider_option('auth_password') is not None)
+        self._zone_data = None
+
+    def zone_data(self):
```
[the remark this answers is not in the document]
Comment (abligh) [reply lexicon/providers/ultradns.py:57]: Fixed

File: lexicon/providers/ultradns.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+            result = requests.post(url, data=data)
+            result.raise_for_status()
+            self.api_key = result.json()['accessToken']
+
+        assert self.api_key is not None
+        self.zone_data()
+        self.domain_id = self.domain
+
+    def _make_identifier(self, rtype, name, content):  # pylint: disable=no-self-use
+        return "{}/{}={}".format(rtype, name, content)
```
[the remark this answers is not in the document]
Comment (abligh) [reply lexicon/providers/ultradns.py]: Fixed

File: lexicon/tests/providers/test_ultradns.py (hunk 1 above)
Comment (abligh) [reply lexicon/tests/providers/test_ultradns.py]: Fixed

File: lexicon/providers/ultradns.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+    def _parse_identifier(self, identifier):  # pylint: disable=no-self-use
```
[the remark this answers is not in the document]
Comment (abligh) [reply lexicon/providers/ultradns.py]: Fixed

**Comment by abligh** [comment]
I think that's all the changes made, and the ci/circleci build error is bogus (it's complaining there is no config file). PTAL!

**Pull request closed without merging**